File: src/genetic_variances.cc
#include "genetic_variances.h"

namespace {
std::atomic_size_t ind;
}

namespace snplib {
void CalcUniLMMThread(const double *traits, const double *covariates,
                      const double *lambda, size_t num_samples,
                      size_t num_covariates, size_t num_traits, double *vars,
                      double *res) {
  UniLMM worker(lambda, covariates, num_samples, num_covariates);
  LineSearch<UniLMM> searcher(worker);
  size_t local_ind = ind++;
  while (local_ind < num_traits) {
    worker.CalcInitialGuess(traits + local_ind * num_samples);
    double a;
    double f_old = worker.CalcLikelihood();
    for (size_t i = 0; i < 15; ++i) {
      worker.CalcEMStep();
      f_old = worker.CalcLikelihood();
    }
    worker.UpdateGradients();
    double f_new = f_old;
    for (size_t l = 0; l < 200; ++l) {
      f_old = f_new;
      worker.BackupVars();
      worker.UpdateHessian();
      worker.CalcAIStep();
      a = searcher.Search(f_old);
      f_new = searcher.GetFNew();
      if ((f_old - f_new) < 1e-8 * (1.0 + std::fabs(f_old))) {
        break;
      }
    }
    auto var = worker.GetVars();
    vars[2 * local_ind] = var[0];
    vars[2 * local_ind + 1] = var[1];
    worker.CalcRes(res + local_ind * num_samples);
    local_ind = ind++;
  }
}
void CalcUniLMM(const double *traits, const double *covariates,
                const double *lambda, size_t num_samples, size_t num_covariates,
                size_t num_traits, double *vars, double *res,
                size_t num_threads) {
  std::vector<std::thread> workers;
  set_num_threads(1);
  ind = 0;
  for (size_t i = 0; i < num_threads; ++i) {
    workers.emplace_back(CalcUniLMMThread, traits, covariates, lambda,
                         num_samples, num_covariates, num_traits, vars, res);
  }
  for (auto &&iter : workers) {
    iter.join();
  }
}

void CalcMultiLMM_REMLThread(const double *covariates, const double *lambda,
                             const double *traits, const double *res,
                             const double *vars, size_t num_samples,
                             size_t num_covariates, size_t num_dims,
                             size_t num_traits, double *sigma_e,
                             double *sigma_g) {
  MultiLMM_REML worker(lambda, covariates, num_samples, num_covariates,
                       num_dims);
  LineSearch<MultiLMM_REML> searcher(worker);
  auto local_ind = ind++;
  while (local_ind < num_traits) {
    worker.CalcInitialGuess(traits + local_ind * num_dims * num_samples,
                            res + local_ind * num_dims * num_samples,
                            vars + local_ind * 2 * num_dims);
    double a;
    double f_old = worker.CalcLikelihood();
    for (size_t i = 0; i < 15; ++i) {
      worker.CalcEMStep();
      f_old = worker.CalcLikelihood();
    }
    worker.UpdateGradients();
    double f_new = f_old;
    for (size_t l = 0; l < 50 * num_dims * num_dims; ++l) {
      f_old = f_new;
      worker.BackupVars();
      worker.UpdateHessian();
      worker.CalcAIStep();
      a = searcher.Search(f_old);
      f_new = searcher.GetFNew();
      if ((f_old - f_new) < 1e-8 * (1.0 + std::fabs(f_old))) {
        break;
      }
    }
    worker.GetSigmaE(sigma_e + local_ind * num_dims * num_dims);
    worker.GetSigmaG(sigma_g + local_ind * num_dims * num_dims);
    local_ind = ind++;
  }
}
void CalcMultiLMM_REML(const double *covariates, const double *lambda,
                       const double *traits, const double *res,
                       const double *vars, size_t num_samples,
                       size_t num_covariates, size_t num_dims,
                       size_t num_traits, double *sigma_e, double *sigma_g,
                       size_t num_threads) {
  std::vector<std::thread> workers;
  set_num_threads(1);
  ind = 0;
  for (size_t i = 0; i < num_threads; ++i) {
    workers.emplace_back(CalcMultiLMM_REMLThread, covariates, lambda, traits,
                         res, vars, num_samples, num_covariates, num_dims,
                         num_traits, sigma_e, sigma_g);
  }
  for (auto &&iter : workers) {
    iter.join();
  }
}
void CalcMultiLMM_RMLThread(const double *lambda, const double *res,
                            const double *vars, size_t num_samples,
                            size_t num_covariates, size_t num_dims,
                            size_t num_traits, double *sigma_e,
                            double *sigma_g) {
  MultiLMM_RML worker(lambda, num_samples, num_covariates, num_dims);
  LineSearch<MultiLMM_RML> searcher(worker);
  auto local_ind = ind++;
  while (local_ind < num_traits) {
    worker.CalcInitialGuess(res + local_ind * num_dims * num_samples,
                            vars + local_ind * 2 * num_dims);
    double a;
    double f_old = worker.CalcLikelihood();
    for (size_t i = 0; i < 15; ++i) {
      worker.CalcEMStep();
      f_old = worker.CalcLikelihood();
    }
    worker.UpdateGradients();
    double f_new = f_old;
    for (size_t l = 0; l < 50 * num_dims * num_dims; ++l) {
      f_old = f_new;
      worker.BackupVars();
      worker.UpdateHessian();
      worker.CalcAIStep();
      a = searcher.Search(f_old);
      f_new = searcher.GetFNew();
      if ((f_old - f_new) < 1e-8 * (1.0 + std::fabs(f_old))) {
        break;
      }
    }
    worker.GetSigmaE(sigma_e + local_ind * num_dims * num_dims);
    worker.GetSigmaG(sigma_g + local_ind * num_dims * num_dims);
    local_ind = ind++;
  }
}

void CalcMultiLMM_RML(const double *lambda, const double *res,
                      const double *vars, size_t num_samples,
                      size_t num_covariates, size_t num_dims, size_t num_traits,
                      double *sigma_e, double *sigma_g, size_t num_threads) {
  std::vector<std::thread> workers;
  ind = 0;
  for (size_t i = 0; i < num_threads; ++i) {
    workers.emplace_back(CalcMultiLMM_RMLThread, lambda, res, vars, num_samples,
                         num_covariates, num_dims, num_traits, sigma_e,
                         sigma_g);
  }
  for (auto &&iter : workers) {
    iter.join();
  }
}
}  // namespace snplib

File: SNPLIB/genetic_variances.py
import math
import re
import numpy as np
import numpy.linalg as npl
from scipy import stats
import _SNPLIB as lib


def CalcUniLMM(traits, covariates, relationship_matrix, num_threads):
    w, v = npl.eig(relationship_matrix)
    std = np.std(traits, axis=0, dtype='double', ddof=1)
    Y = v.T@stats.zscore(traits, axis=0, ddof=1)
    X = v.T@covariates
    Vars, Res = lib.CalcUniLMM(Y, X, w, num_threads)
    Vars = Vars*(std**2)
    Res = Res*std
    return Vars, Res


def CalcMultiLMM_REML(traits, covariates, relationship_matrix, num_dims, num_threads):
    w, v = npl.eig(relationship_matrix)
    std = np.std(traits, axis=0, dtype='double', ddof=1)
    Y = v.T@stats.zscore(traits, axis=0, ddof=1)
    X = v.T@covariates
    Vars, Res = lib.CalcUniLMM(Y, X, w, num_threads)
    sigma_e, sigma_g = lib.CalcMultiLMM_REML(
        Y, X, w, Res, Vars, num_dims, num_threads)
    num_traits = Y.shape[1]//num_dims
    for i in range(num_traits):
        D = np.diag(std[i*num_dims:(i+1)*num_dims])
        sigma_e[..., i] = D@sigma_e[..., i]@D
        sigma_g[..., i] = D@sigma_g[..., i]@D
    return sigma_e, sigma_g


def CalcMultiLMM_RML(traits, covariates, relationship_matrix, num_dims, num_threads):
    w, v = npl.eig(relationship_matrix)
    std = np.std(traits, axis=0, dtype='double', ddof=1)
    Y = v.T@stats.zscore(traits, axis=0, ddof=1)
    X = v.T@covariates
    Vars, Res = lib.CalcUniLMM(Y, X, w, num_threads)
    sigma_e, sigma_g = lib.CalcMultiLMM_RML(
        w, Res, Vars, num_dims, num_threads)
    num_traits = Y.shape[1]//num_dims
    for i in range(num_traits):
        D = np.diag(std[i*num_dims:(i+1)*num_dims])
        sigma_e[..., i] = D@sigma_e[..., i]@D
        sigma_g[..., i] = D@sigma_g[..., i]@D
    return sigma_e, sigma_g
